sys/kern/interrupt.c

```diff
@@ -157,30 +183,61 @@ void intr_event_run_handlers(intr_event_t *ie) {
 }
 
 static void intr_thread(void *arg) {
+  intr_thread_t *it = (intr_thread_t *)arg;
+
   while (true) {
-    intr_handler_t *ih;
+    intr_handler_t *ih, *next, *curr;
+
+    ih = NULL;
 
+    /* We look at the list of handlers of our event and search for a handler
+     * marked as IH_DELEGATE. If we don't find any handlers to be serviced, we
+     * go to sleep on our wchan.
+     */
     WITH_INTR_DISABLED {
-      while (TAILQ_EMPTY(&delegated))
-        sleepq_wait(&delegated, NULL);
-      ih = TAILQ_FIRST(&delegated);
-      TAILQ_REMOVE(&delegated, ih, ih_link);
+      TAILQ_FOREACH_SAFE (curr, &it->it_event->ie_handlers, ih_link, next) {
+        if (curr->ih_flags & IH_DELEGATE) {
+          ih = curr;
+          break;
+        }
+      }
+      if (ih == NULL)
+        sleepq_wait(&it->it_wchan, NULL);
     }
 
     ih->ih_service(ih->ih_argument);
```

Comment:
Wait a minute. Do I understand it correctly? I think that's what you do:
1. You look for a handler within an interrupt event which needs to be serviced.
2. Now, let's say you won't find any (i.e. `ih == NULL`).
3. You go to sleep on the wait channel.
4. You've reached line 208, and... `ih` is stlii NULL.

If I am right, what you need to do is to put in to a `do {} while (it == NULL);` loop.